libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs

```diff
@@ -894,6 +895,114 @@ public virtual async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext contex
+            var tokenExchangeState = new TokenExchangeState()
+            {
+                ConnectionName = connectionName,
+                Conversation = new ConversationReference()
+                {
+                    ActivityId = activity.Id,
+                    Bot = activity.Recipient,       // Activity is from the user to the bot
+                    ChannelId = activity.ChannelId,
+                    Conversation = activity.Conversation,
+                    ServiceUrl = activity.ServiceUrl,
+                    User = activity.From,
+                },
+                MsAppId = appId,
+            };
+
+            var serializedState = JsonConvert.SerializeObject(tokenExchangeState);
+            var encodedState = Encoding.UTF8.GetBytes(serializedState);
+            var state = Convert.ToBase64String(encodedState);
+
+            var client = await CreateOAuthApiClientAsync(turnContext).ConfigureAwait(false);
+
+            return await client.BotSignIn.GetSignInResourceAsync(state, null, null, finalRedirect, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Performs a token exchange operation such as for single sign-on.
+        /// </summary>
+        /// <param name="turnContext">Context for the current turn of conversation with the user.</param>
+        /// <param name="connectionName">Name of the auth connection to use.</param>
+        /// <param name="userId">The user id associated with the token..</param>
+        /// <param name="exchangeRequest">The exchange request details, either a token to exchange or a uri to exchange.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>If the task completes, the exchanged token is returned.</returns>
+        public virtual async Task<TokenResponse> ExchangeTokenAsync(ITurnContext turnContext, string connectionName, string userId, TokenExchangeRequest exchangeRequest, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var client = await CreateOAuthApiClientAsync(turnContext).ConfigureAwait(false);
+            var result = await client.UserToken.ExchangeAsyncAsync(userId, connectionName, turnContext.Activity.ChannelId, exchangeRequest, cancellationToken).ConfigureAwait(false);
+
+            if (result is ErrorResponse errorResponse)
+            {
+                result = new TokenResponse();
+            }
+
+            if (result is TokenResponse tokenResponse)
+            {
+                // todo: temporary
+                if (string.IsNullOrEmpty(tokenResponse.Token))
+                {
+                    if (!string.IsNullOrEmpty(exchangeRequest.Uri))
+                    {
+                        tokenResponse.Token = "exchangeableToken" + Guid.NewGuid().ToString().Replace("-", string.Empty);
```

Comment:
Guid.NewGuid().ToString("N");

Reply:
This code should be removed.

---
In reply to: [375491057](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375491057) [](ancestors = 375491057)